components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/CustomUserSelfRegistrationHandler.java

```diff
@@ -0,0 +1,101 @@
+
+        //The handler should be called ss a post add user event.
+        if (IdentityEventConstants.Event.POST_ADD_USER.equals(event.getEventName())) {
+            try {
+                addNewRole(tenantDomain, userName);
+            } catch (IdentityRecoveryServerException e) {
+                throw new IdentityEventException("Error while adding custom roles to the user", e);
+            }
+        }
+    }
+
+    private void addNewRole(String tenantDomain, String userName)
+            throws org.wso2.carbon.identity.recovery.IdentityRecoveryServerException {
+        try {
+            //Realm service is used for user management tasks
+            RealmService realmService = APIKeyMgtDataHolder.getRealmService();
+            UserStoreManager userStoreManager;
+            try {
+                userStoreManager = realmService.getTenantUserRealm(IdentityTenantUtil.getTenantId(tenantDomain))
+                        .getUserStoreManager();
+            } catch (UserStoreException e) {
+                throw Utils
+                        .handleServerException(IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_UNEXPECTED, userName,
+                                e);
+            }
+            //Start a tenant flow
+            PrivilegedCarbonContext.startTenantFlow();
+            PrivilegedCarbonContext carbonContext = PrivilegedCarbonContext.getThreadLocalCarbonContext();
+            carbonContext.setTenantId(IdentityTenantUtil.getTenantId(tenantDomain));
+            carbonContext.setTenantDomain(tenantDomain);
+            try {
+                //Since this handler is called as a post add user event, the user should exists in the userstore
+                if (userStoreManager.isExistingUser(userName)) {
+                    List<String> roleList = asList(userStoreManager.getRoleListOfUser(userName));
+                    //User should have selfSignup role. Checking whether the user is in the new role
+                    if (roleList.contains(SELF_SIGNUP_ROLE) && !roleList.contains(SUBSCRIBER_ROLE)) {
+                        String[] userRoles = new String[] { SUBSCRIBER_ROLE };
+                        userStoreManager.updateRoleListOfUser(userName, null, userRoles);
+                    }
+                }
+            } catch (UserStoreException e) {
+                throw Utils
+                        .handleServerException(IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_UNEXPECTED, userName,
+                                e);
+            }
+        } finally {
+            Utils.clearArbitraryProperties();
+            PrivilegedCarbonContext.endTenantFlow();
+        }
+    }
+
+    @Override
+    public void init(InitConfig configuration) throws IdentityRuntimeException {
+        super.init(configuration);
+    }
+
+    @Override
+    public int getPriority(MessageContext messageContext) {
+        return 250;
```

Comment:
How does this number decided? Please use a meaningful constant.

Reply:
High value means priority is low. that  means the handler gets executed at last. There is no reason to use 250. just wanted to give it a high value so that this will get executed last.

Reply:
Ok.
let's use a constant with a meaningful name.

Reply:
resolved